### PR TITLE
#52 Contribute BDD operations `relnext` and `relprev` and their restricted variants

### LINKED ISSUE #52: Add support for `relprev` and `relnext` (and variants)
For symbolic reachability there exist specialized operations `relnext` and `relprev`, that compute the set of states reachable by taking one step forward/backward with respect to a given transition relation. Such operations are more efficient than computing `applyEx` followed by `replace`.

There are various variants of `relnext` and `relprev` that may further increase efficiency. For example, `relnextRestricted` and `relprevRestricted` can compute `and(relnext(...), restriction)` in a single step with respect to some BDD `restriction`, and likewise for `relprevRestricted`.

Goal of this issue is contributing `relnext`, `relprev`, `relnextRestricted`, and `relprevRestricted` to JavaBDD.

### PR DESCRIPTION
**Pull request by wytseoortwijn**
Closes #52

### CODE REVIEW

File: src/main/java/com/github/javabdd/BDD.java [hunk 1]
```diff
@@ -586,6 +586,56 @@ public BDD relprod(BDD that, BDDVarSet var) {
      */
     public abstract BDD satOne(BDDVarSet var, boolean pol);
 
+    /**
+     * Computes the BDD that represents the set of successor states from {@code states}, i.e., the ones reachable by
+     * taking one forward step from states in {@code states}, where this BDD is the (partial) transition relation to
+     * consider. It is assumed that old-state and new-state BDD variables are interleaved in this transition relation.
```
Comment (dhendriks) [inline src/main/java/com/github/javabdd/BDD.java:592]: It indicates interleaving is required, but not that the old var needs to be before the new var, which is also assumed in the implementation.

File: src/main/java/com/github/javabdd/BDD.java [hunk 2]
```diff
@@ -586,6 +586,56 @@ public BDD relprod(BDD that, BDDVarSet var) {
...
+     *
+     * @param states The BDD representing the set of states.
+     * @param vars The BDD representing the set of variables that are relevant to consider for determining successor
+     * states. This variable set <i>must</i> include all variables that occur in the transition relation BDD. Moreover,
+     * for every new-state variable that is in {@code vars}, {@code relnext} considers the corresponding old-state
+     * variable to be relevant as well, regardless of whether it's included in {@code vars} or not.
```
Comment (dhendriks) [inline src/main/java/com/github/javabdd/BDD.java:598]: Does it not consider for old vars also the new vars? I think it does.

File: src/main/java/com/github/javabdd/BDDFactoryIntImpl.java [hunk 3]
```diff
@@ -86,6 +86,16 @@ public abstract class BDDFactoryIntImpl extends BDDFactory {
...
     protected abstract /* bdd */int veccompose_impl(/* bdd */int v, BDDPairing p);
 
+    protected abstract /* bdd */int relnext_impl(/* bdd */int states, /* bdd */int relation, /* bdd */int vars);
+
+    protected abstract /* bdd */int relnextRestricted_impl(/* bdd */int states, /* bdd */int relation,
+            /* bdd */int restriction, /* bdd */int vars);
+
+    protected abstract /* bdd */int relprev_impl(/* bdd */int relation, /* bdd */int states, /* bdd */int vars);
+
+    protected abstract /* bdd */int relprevRestricted_impl(/* bdd */int relation, /* bdd */int states,
```
Comment (dhendriks) [inline src/main/java/com/github/javabdd/BDDFactoryIntImpl.java:96]: Is there a reason the `states` and `relation` arguments are swapped between `relprev*` and `relnext*`? Not really a problem, just curious.

File: src/main/java/com/github/javabdd/JFactory.java [hunk 4]
```diff
@@ -1497,8 +1541,8 @@ int ite_rec(int f, int g, int h) {
             return not_rec(f);
         }
 
-        entry = BddCache_lookupI(itecache, ITEHASH(f, g, h));
```
Comment (dhendriks) [inline src/main/java/com/github/javabdd/JFactory.java:1500]: Is ITEHASH still used? Should it be removed?

File: src/main/java/com/github/javabdd/JFactory.java [hunk 5]
```diff
@@ -2377,6 +2425,974 @@ int zdiff_rec(int l, int r) {
...
+    int relnext_rec(int states, int relation, int vars) {
+        if (VERIFY_ASSERTIONS) {
+            _assert(!ZDD);
+        }
+
+        if (cachestats.enabled) {
+            cachestats.opAccess++;
+        }
+
+        // Terminals cases.
```
Comment (dhendriks) [inline src/main/java/com/github/javabdd/JFactory.java:2482]: what if states is not zero/one, and relation is one, then result is one, right? probably won't occur in practice though?

File: src/main/java/com/github/javabdd/JFactory.java [hunk 6]
```diff
@@ -2377,6 +2425,974 @@ int zdiff_rec(int l, int r) {
...
+            return entry.res;
+        }
+
+        if (cachestats.enabled) {
+            cachestats.opMiss++;
+        }
+
+        int result;
+
+        if (sameHeight) {
```
Comment (dhendriks) [inline src/main/java/com/github/javabdd/JFactory.java:2536]: from here this method is difficult to review for me...

File: src/main/java/com/github/javabdd/JFactory.java [hunk 7]
```diff
@@ -2377,6 +2425,974 @@ int zdiff_rec(int l, int r) {
...
+            }
+
+            vars = HIGH(vars);
+
+            if (ISCONST(vars)) {
+                return states;
+            }
+        }
+
+        // Consult the operation cache.
```
Comment (dhendriks) [inline src/main/java/com/github/javabdd/JFactory.java:2520]: missing section comment for done with consulting cache, but now doing the next thing (the actual operation?).

File: src/main/java/com/github/javabdd/JFactory.java [hunk 8]
```diff
@@ -2377,6 +2425,974 @@ int zdiff_rec(int l, int r) {
...
+            }
+            if (level_restriction == level_oldvar) {
+                u0 = LOW(restriction);
+                u1 = HIGH(restriction);
+            } else {
+                u0 = restriction;
+                u1 = restriction;
+            }
+
+            // Update the variable set, and while doing so, determine whether the new state variable must be considered.
```
Comment (dhendriks) [inline src/main/java/com/github/javabdd/JFactory.java:3303]: why does only `relprevRestricted_rec` have this comment, and not the other 3?

File: src/main/java/com/github/javabdd/BDD.java (hunk 2 above)
Comment (wytseoortwijn) [reply src/main/java/com/github/javabdd/BDD.java:598]: If some old-state var `x` is in `vars` but the corresponding new-state variable `x'` is not, then `relnext`/`relprev`/... will not apply quantification and thus the value of `x` will not change.

If the new-state variable `x'` happens to be in `vars` but the old-state variable `x` is not, then still `relnext`/`relprev`/... may consider different values for `x`, hence it must be considered relevant. (If we want we could turn this into a precondition, and require that if `x'` is in `vars` then also `x` must be in `vars`. But the implementation now works fine also without this assumption. It just considers `x` relevant if `x'` is relevant.)

File: src/main/java/com/github/javabdd/BDDFactoryIntImpl.java (hunk 3 above)
Comment (wytseoortwijn) [reply src/main/java/com/github/javabdd/BDDFactoryIntImpl.java:96]: This is taken over from Sylvan like this. In the `BDD` class `relnext` and `relprev` have the same signature. So users of JavaBDD would not really notice.

File: src/main/java/com/github/javabdd/JFactory.java (hunk 5 above)
Comment (wytseoortwijn) [reply src/main/java/com/github/javabdd/JFactory.java:2482]: That's only true if all leftover variables to consider in `states` are allowed to change. That is, it's only true if for every old-state BDD variable `x` that still occurs in `states`, it holds that the corresponding new-state variable `x'` is in `vars`.

**Comment by wytseoortwijn** [comment]
Ready for next review.

File: src/main/java/com/github/javabdd/BDD.java (hunk 2 above)
Comment (dhendriks) [reply src/main/java/com/github/javabdd/BDD.java:599]: OK, clear. I think we can leave it as is. It is properly documented in the JavaDoc.

Comment (dhendriks) [reply src/main/java/com/github/javabdd/BDD.java:599]: Maybe one more question: why does this not differ for the different directions (`relprev`/`relnext`)?

File: src/main/java/com/github/javabdd/JFactory.java (hunk 5 above)
Comment (dhendriks) [reply src/main/java/com/github/javabdd/JFactory.java:2478]: Ah, right. Never mind then.

File: src/main/java/com/github/javabdd/BDDFactoryIntImpl.java (hunk 3 above)
Comment (dhendriks) [reply src/main/java/com/github/javabdd/BDDFactoryIntImpl.java:96]: Indeed, I saw it was internal. Thanks for the clarification. Let's keep it as is.

File: src/main/java/com/github/javabdd/BDD.java [hunk 9]
```diff
@@ -586,6 +586,58 @@ public BDD relprod(BDD that, BDDVarSet var) {
      */
     public abstract BDD satOne(BDDVarSet var, boolean pol);
 
+    /**
+     * Computes the BDD that represents the set of successor states from {@code states}, i.e., the ones reachable by
+     * taking one forward step from states in {@code states}, where this BDD is the (partial) transition relation to
+     * consider. It is assumed that old-state and new-state BDD variables are interleaved in this transition relation,
+     * starting with the old-state variable, i.e., the old-state variable comes before the new-state variable.
```
Comment (dhendriks) [inline src/main/java/com/github/javabdd/BDD.java:593]:
```suggestion
     * starting with the old-state variable, i.e., each old-state variable comes directly before its corresponding new-state variable.
```

File: src/main/java/com/github/javabdd/BDD.java [hunk 10]
```diff
@@ -586,6 +586,58 @@ public BDD relprod(BDD that, BDDVarSet var) {
...
+     * relnext} for further details.
+     * @return The BDD representing the restricted set of successor states from {@code states}.
+     */
+    public abstract BDD relnextRestricted(BDD states, BDD restriction, BDDVarSet vars);
+
+    /**
+     * Computes the BDD that represents the set of predecessor states from {@code states}, i.e., the ones reachable by
+     * taking one backward step from states in {@code states}, where this BDD is the (partial) transition relation to
+     * consider. It is assumed that old-state and new-state BDD variables are interleaved in this transition relation,
+     * starting with the old-state variable, i.e., the old-state variable comes before the new-state variable.
```
Comment (dhendriks) [inline src/main/java/com/github/javabdd/BDD.java:619]:
```suggestion
     * starting with the old-state variable, i.e., each old-state variable comes directly before its corresponding new-state variable.
```

**Review by dhendriks — CHANGES_REQUESTED** [review]

File: src/main/java/com/github/javabdd/BDD.java (hunk 2 above)
Comment (wytseoortwijn) [reply src/main/java/com/github/javabdd/BDD.java:599]: The transition relation that's given to `relnext(..., relation, ...)`/`relprev(..., relation, ...)`/... stays the same regardless of the direction in which it is applied. In both cases, `relation` may e.g. quantify over some new-state variables, and then it's up to `relnext`/`relprev`/... to handle that correctly.

**Comment by wytseoortwijn** [comment]
Ready for next review

File: src/main/java/com/github/javabdd/BDD.java [hunk 11]
```diff
@@ -586,6 +586,58 @@ public BDD relprod(BDD that, BDDVarSet var) {
      */
     public abstract BDD satOne(BDDVarSet var, boolean pol);
 
+    /**
+     * Computes the BDD that represents the set of successor states from {@code states}, i.e., the ones reachable by
+     * taking one forward step from states in {@code states}, where this BDD is the (partial) transition relation to
+     * consider. It is assumed that old-state and new-state BDD variables are interleaved in this transition relation,
+     * starting with the old-state variable, i.e., each old-state variable comes directly before its corresponding new-state variable.
```
Comment (dhendriks) [inline src/main/java/com/github/javabdd/BDD.java:593]: File requires reformat/rewrap.

**Review by dhendriks — CHANGES_REQUESTED** [review]

**Review by dhendriks — APPROVED** [review]

**Pull request merged**